File: litho-it/src/test/java/com/facebook/litho/IncrementalMountExtensionTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho;

import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import android.graphics.Rect;
import androidx.annotation.Nullable;
import com.facebook.litho.config.ComponentsConfiguration;
import com.facebook.litho.testing.testrunner.LithoTestRunner;
import com.facebook.rendercore.MountDelegate;
import com.facebook.rendercore.MountDelegateInput;
import com.facebook.rendercore.MountDelegateTarget;
import com.facebook.rendercore.RenderTreeNode;
import com.facebook.rendercore.RenderUnit;
import com.facebook.rendercore.extensions.ExtensionState;
import com.facebook.rendercore.incrementalmount.IncrementalMountExtension;
import com.facebook.rendercore.incrementalmount.IncrementalMountExtension.IncrementalMountExtensionState;
import com.facebook.rendercore.incrementalmount.IncrementalMountExtensionInput;
import com.facebook.rendercore.incrementalmount.IncrementalMountOutput;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.ParameterizedRobolectricTestRunner;

@RunWith(LithoTestRunner.class)
public class IncrementalMountExtensionTest {

  @Test
  public void onIteratingOnIncrementMountOutputs_shouldIterateByInsertionOrder() {
    final TestInput input = new TestInput(10);
    int i = 0;
    for (IncrementalMountOutput output : input.getIncrementalMountOutputs()) {
      assertThat(output.getId()).isEqualTo(input.getMountableOutputAt(i).getRenderUnit().getId());
      i++;
    }
  }

  @Test
  public void testDirtyMountWithEmptyRect() {
    final LithoView lithoView = mock(LithoView.class);
    when(lithoView.getHeight()).thenReturn(50);
    final MountDelegate mountDelegate = mock(MountDelegate.class);
    final MountDelegateTarget mountDelegateTarget = mock(MountDelegateTarget.class);
    when(mountDelegate.getMountDelegateTarget()).thenReturn(mountDelegateTarget);
    final IncrementalMountExtension extension = IncrementalMountExtension.getInstance();

    final ExtensionState<IncrementalMountExtensionState> extensionState =
        extension.createExtensionState(mountDelegate);
    final IncrementalMountExtensionState state = extensionState.getState();

    mountDelegate.addExtension(extension);
    when(mountDelegate.getExtensionState(extension)).thenReturn(extensionState);

    final TestInput incrementalMountExtensionInput = new TestInput(10);

    extension.beforeMount(extensionState, incrementalMountExtensionInput, new Rect(0, 0, 10, 50));
    for (int i = 0, size = incrementalMountExtensionInput.getMountableOutputCount();
        i < size;
        i++) {
      final RenderTreeNode node = incrementalMountExtensionInput.getMountableOutputAt(i);
      extension.beforeMountItem(extensionState, node, i);
    }
    extension.afterMount(extensionState);

    assertThat(extension.getPreviousBottomsIndex(state)).isEqualTo(0);
    assertThat(extension.getPreviousTopsIndex(state)).isEqualTo(5);

    final TestInput incrementalMountExtensionInput2 = new TestInput(3);
    extension.beforeMount(extensionState, incrementalMountExtensionInput2, new Rect(0, 0, 0, 0));
    for (int i = 0, size = incrementalMountExtensionInput2.getMountableOutputCount();
        i < size;
        i++) {
      final RenderTreeNode node = incrementalMountExtensionInput2.getMountableOutputAt(i);
      extension.beforeMountItem(extensionState, node, i);
    }
    extension.afterMount(extensionState);

    extension.onVisibleBoundsChanged(extensionState, new Rect(0, 0, 10, 50));

    assertThat(extension.getPreviousBottomsIndex(state)).isEqualTo(0);
    assertThat(extension.getPreviousTopsIndex(state)).isEqualTo(3);
  }

  @Test
  public void testDirtyMountWithEmptyRect_leftRightMatch() {
    final LithoView lithoView = mock(LithoView.class);
    when(lithoView.getHeight()).thenReturn(50);
    final MountDelegate mountDelegate = mock(MountDelegate.class);
    final MountDelegateTarget mountDelegateTarget = mock(MountDelegateTarget.class);
    when(mountDelegate.getMountDelegateTarget()).thenReturn(mountDelegateTarget);
    final IncrementalMountExtension extension = IncrementalMountExtension.getInstance();
    mountDelegate.addExtension(extension);

    final ExtensionState<IncrementalMountExtensionState> extensionState =
        extension.createExtensionState(mountDelegate);
    final IncrementalMountExtensionState state = extensionState.getState();
    mountDelegate.addExtension(extension);
    when(mountDelegate.getExtensionState(extension)).thenReturn(extensionState);

    final TestInput incrementalMountExtensionInput = new TestInput(10);

    extension.beforeMount(extensionState, incrementalMountExtensionInput, new Rect(0, 0, 10, 50));
    for (int i = 0, size = incrementalMountExtensionInput.getMountableOutputCount();
        i < size;
        i++) {
      final RenderTreeNode node = incrementalMountExtensionInput.getMountableOutputAt(i);
      extension.beforeMountItem(extensionState, node, i);
    }
    extension.afterMount(extensionState);

    assertThat(extension.getPreviousBottomsIndex(state)).isEqualTo(0);
    assertThat(extension.getPreviousTopsIndex(state)).isEqualTo(5);

    final TestInput incrementalMountExtensionInput2 = new TestInput(3);
    extension.beforeMount(extensionState, incrementalMountExtensionInput2, new Rect(0, 0, 10, 0));
    for (int i = 0, size = incrementalMountExtensionInput2.getMountableOutputCount();
        i < size;
        i++) {
      final RenderTreeNode node = incrementalMountExtensionInput2.getMountableOutputAt(i);
      extension.beforeMountItem(extensionState, node, i);
    }
    extension.afterMount(extensionState);

    extension.onVisibleBoundsChanged(extensionState, new Rect(0, 0, 10, 50));

    assertThat(extension.getPreviousBottomsIndex(state)).isEqualTo(0);
    assertThat(extension.getPreviousTopsIndex(state)).isEqualTo(3);
  }

  final class TestInput implements IncrementalMountExtensionInput, MountDelegateInput {
    final List<RenderTreeNode> mountableOutputs = new ArrayList<>();
    final Map<Long, IncrementalMountOutput> mIncrementalMountOutputs = new LinkedHashMap<>();
    final List<IncrementalMountOutput> tops = new ArrayList<>();
    final List<IncrementalMountOutput> bottoms = new ArrayList<>();
    private final int mCount;

    public TestInput(int count) {
      mCount = count;
      for (int i = 0; i < count; i++) {
        final Rect bounds = new Rect(0, (i * 10), 10, (i + 1) * 10);

        final LayoutOutput layoutOutput = mock(LayoutOutput.class);
        RenderTreeNode renderTreeNode = mock(RenderTreeNode.class);
        when(renderTreeNode.getLayoutData()).thenReturn(layoutOutput);
        when(renderTreeNode.getAbsoluteBounds(any(Rect.class))).thenReturn(bounds);
        when(layoutOutput.getComponent()).thenReturn(mock(Component.class));
        when(layoutOutput.getBounds()).thenReturn(bounds);

        RenderUnit renderUnit = new LithoRenderUnit(layoutOutput);
        when(renderUnit.getId()).thenReturn((long) i);
        when(renderTreeNode.getRenderUnit()).thenReturn(renderUnit);

        mountableOutputs.add(renderTreeNode);
        final IncrementalMountOutput incrementalMountOutput =
            new IncrementalMountOutput(
                i, i, bounds, i != 0 ? mIncrementalMountOutputs.get((long) (i - 1)) : null);
        mIncrementalMountOutputs.put(incrementalMountOutput.getId(), incrementalMountOutput);
        tops.add(incrementalMountOutput);
        bottoms.add(incrementalMountOutput);
      }
    }

    @Override
    public int getMountableOutputCount() {
      return mCount;
    }

    @Override
    public List<IncrementalMountOutput> getOutputsOrderedByTopBounds() {
      return tops;
    }

    @Override
    public List<IncrementalMountOutput> getOutputsOrderedByBottomBounds() {
      return bottoms;
    }

    @Override
    public @Nullable IncrementalMountOutput getIncrementalMountOutputForId(long id) {
      return mIncrementalMountOutputs.get(id);
    }

    @Override
    public Collection<IncrementalMountOutput> getIncrementalMountOutputs() {
      return mIncrementalMountOutputs.values();
    }

    @Override
    public int getIncrementalMountOutputCount() {
      return mCount;
    }

    @Override
    public int getPositionForId(long id) {
      return 0;
    }

    @Override
    public boolean renderUnitWithIdHostsRenderTrees(long id) {
      return true;
    }

    @Override
    public RenderTreeNode getMountableOutputAt(int position) {
      return mountableOutputs.get(position);
    }
  }
}

File: litho-it/src/test/java/com/facebook/litho/IncrementalMountExtensionTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho

import android.graphics.Color
import android.graphics.Rect
import android.graphics.drawable.ColorDrawable
import com.facebook.litho.testing.testrunner.LithoTestRunner
import com.facebook.rendercore.MountDelegate
import com.facebook.rendercore.MountDelegateInput
import com.facebook.rendercore.MountDelegateTarget
import com.facebook.rendercore.RenderCoreSystrace
import com.facebook.rendercore.RenderTreeNode
import com.facebook.rendercore.RenderUnit
import com.facebook.rendercore.incrementalmount.IncrementalMountExtension
import com.facebook.rendercore.incrementalmount.IncrementalMountExtensionInput
import com.facebook.rendercore.incrementalmount.IncrementalMountOutput
import com.facebook.rendercore.testing.DrawableWrapperUnit
import java.util.ArrayList
import java.util.LinkedHashMap
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

@RunWith(LithoTestRunner::class)
class IncrementalMountExtensionTest {

  @Test
  fun onIteratingOnIncrementMountOutputs_shouldIterateByInsertionOrder() {
    val input = TestInput(10)
    input.incrementalMountOutputs.forEachIndexed { i, output ->
      assertThat(output.id).isEqualTo(input.getMountableOutputAt(i).renderUnit.id)
    }
  }

  @Test
  fun testDirtyMountWithEmptyRect() {
    val lithoView = mock<LithoView>()
    whenever(lithoView.height).thenReturn(50)
    val mountDelegate = mock<MountDelegate>()
    val mountDelegateTarget = mock<MountDelegateTarget>()
    whenever(mountDelegate.tracer).thenReturn(RenderCoreSystrace.getInstance())
    whenever(mountDelegate.mountDelegateTarget).thenReturn(mountDelegateTarget)
    val extension = IncrementalMountExtension.getInstance()
    val extensionState = extension.createExtensionState(mountDelegate)
    val state = extensionState.state
    mountDelegate.registerMountExtension(extension)
    val incrementalMountExtensionInput = TestInput(10)
    extension.beforeMount(extensionState, incrementalMountExtensionInput, Rect(0, 0, 10, 50))
    for (i in 0 until incrementalMountExtensionInput.mountableOutputCount) {
      val node = incrementalMountExtensionInput.getMountableOutputAt(i)
      extension.beforeMountItem(extensionState, node, i)
    }
    extension.afterMount(extensionState)
    assertThat(IncrementalMountExtension.getPreviousBottomsIndex(state)).isEqualTo(0)
    assertThat(IncrementalMountExtension.getPreviousTopsIndex(state)).isEqualTo(5)
    val incrementalMountExtensionInput2 = TestInput(3)
    extension.beforeMount(extensionState, incrementalMountExtensionInput2, Rect(0, 0, 0, 0))
    for (i in 0 until incrementalMountExtensionInput2.mountableOutputCount) {
      val node = incrementalMountExtensionInput2.getMountableOutputAt(i)
      extension.beforeMountItem(extensionState, node, i)
    }
    extension.afterMount(extensionState)
    extension.onVisibleBoundsChanged(extensionState, Rect(0, 0, 10, 50))
    assertThat(IncrementalMountExtension.getPreviousBottomsIndex(state)).isEqualTo(0)
    assertThat(IncrementalMountExtension.getPreviousTopsIndex(state)).isEqualTo(3)
  }

  @Test
  fun testDirtyMountWithEmptyRect_leftRightMatch() {
    val lithoView = mock<LithoView>()
    whenever(lithoView.height).thenReturn(50)
    val mountDelegate = mock<MountDelegate>()
    val mountDelegateTarget = mock<MountDelegateTarget>()
    whenever(mountDelegate.tracer).thenReturn(RenderCoreSystrace.getInstance())
    whenever(mountDelegate.mountDelegateTarget).thenReturn(mountDelegateTarget)
    val extension = IncrementalMountExtension.getInstance()
    mountDelegate.registerMountExtension(extension)
    val extensionState = extension.createExtensionState(mountDelegate)
    val state = extensionState.state
    mountDelegate.registerMountExtension(extension)
    val incrementalMountExtensionInput = TestInput(10)
    extension.beforeMount(extensionState, incrementalMountExtensionInput, Rect(0, 0, 10, 50))
    for (i in 0 until incrementalMountExtensionInput.mountableOutputCount) {
      val node = incrementalMountExtensionInput.getMountableOutputAt(i)
      extension.beforeMountItem(extensionState, node, i)
    }
    extension.afterMount(extensionState)
    assertThat(IncrementalMountExtension.getPreviousBottomsIndex(state)).isEqualTo(0)
    assertThat(IncrementalMountExtension.getPreviousTopsIndex(state)).isEqualTo(5)
    val incrementalMountExtensionInput2 = TestInput(3)
    extension.beforeMount(extensionState, incrementalMountExtensionInput2, Rect(0, 0, 10, 0))
    for (i in 0 until incrementalMountExtensionInput2.mountableOutputCount) {
      val node = incrementalMountExtensionInput2.getMountableOutputAt(i)
      extension.beforeMountItem(extensionState, node, i)
    }
    extension.afterMount(extensionState)
    extension.onVisibleBoundsChanged(extensionState, Rect(0, 0, 10, 50))
    assertThat(IncrementalMountExtension.getPreviousBottomsIndex(state)).isEqualTo(0)
    assertThat(IncrementalMountExtension.getPreviousTopsIndex(state)).isEqualTo(3)
  }

  internal inner class TestInput(private val count: Int) :
      IncrementalMountExtensionInput, MountDelegateInput {
    private val mountableOutputs: MutableList<RenderTreeNode> = ArrayList()
    private val _incrementalMountOutputs: MutableMap<Long, IncrementalMountOutput> = LinkedHashMap()
    private val tops: MutableList<IncrementalMountOutput> = ArrayList()
    private val bottoms: MutableList<IncrementalMountOutput> = ArrayList()

    override fun getMountableOutputCount(): Int = count

    override fun getOutputsOrderedByTopBounds(): List<IncrementalMountOutput> = tops

    override fun getOutputsOrderedByBottomBounds(): List<IncrementalMountOutput> = bottoms

    override fun getIncrementalMountOutputForId(id: Long): IncrementalMountOutput? =
        _incrementalMountOutputs[id]

    override fun getIncrementalMountOutputs(): Collection<IncrementalMountOutput> =
        _incrementalMountOutputs.values

    override fun getIncrementalMountOutputCount(): Int = count

    override fun getPositionForId(id: Long): Int = 0

    override fun renderUnitWithIdHostsRenderTrees(id: Long): Boolean = true

    override fun getMountableOutputAt(position: Int): RenderTreeNode = mountableOutputs[position]

    init {
      for (i in 0 until count) {
        val bounds = Rect(0, i * 10, 10, (i + 1) * 10)
        val renderTreeNode = mock<RenderTreeNode>()
        whenever(renderTreeNode.layoutData).thenReturn(mock<LithoAnimtableItem>())
        whenever(renderTreeNode.getAbsoluteBounds(anyOrNull<Rect>())).thenReturn(bounds)
        val renderUnit: RenderUnit<*> = DrawableWrapperUnit(ColorDrawable(Color.BLACK), i.toLong())
        whenever(renderTreeNode.renderUnit).thenReturn(renderUnit)
        mountableOutputs.add(renderTreeNode)
        val incrementalMountOutput =
            IncrementalMountOutput(
                i.toLong(),
                i,
                bounds,
                false,
                if (i != 0) _incrementalMountOutputs[(i - 1).toLong()] else null)
        _incrementalMountOutputs[incrementalMountOutput.id] = incrementalMountOutput
        tops.add(incrementalMountOutput)
        bottoms.add(incrementalMountOutput)
      }
    }
  }
}
